### DISCUSSION TITLE
Content quality checkers

### DISCUSSION THREAD

**Discussion by benenright** [wagtail/wagtail#11063] [General]
### Summary
Exploring ways in which content quality checkers can be integrated into the wagtail editor experience, and how this could combine with other types of checks such as accessibility checks and sustainability scores. 

### Background
- See https://github.com/wagtail/roadmap/issues/50

### Initial design exploration
- Derived inspiration from Hemmingway and Grammarly.
- Explored data representation: compared pie charts and spark lines for illustrating content quality progression.
- Investigated a combined panel showcasing content quality, sustainability, and accessibility (a11y) checks, featuring an overarching content quality score.
- Considered surfacing average content quality across all pages on the dashboard, spotlighting content requiring enhancement.
- Explored integration potential between automated checks/comments and manual commenting panels.
- Drafted ‘suggestions’ that are moderately assertive to avoid editor frustration, utilizing blue informational alerts instead of more intense red/amber alerts.
- Weighed up the intrusiveness of suggestion mechanisms.
- Explored a toggle for ‘live analysis’ to control real-time issue highlighting during typing.
- Analyzed page weight considerations like adherence to recommended maximum image count for sustainability.
- Considered using icons for varying suggestion types (content quality, sustainability, a11y, etc.), albeit possibly omitting groupings in the MVP.
- Explored interfacing with AI (e.g., ChatGPT) for in-place suggestion and replacement to assist in achieving ATAG compliance.
- For ATAG compliance, examined prompting mechanisms, such as a ‘publish’ modal, encouraging users to utilize content checkers for quality assurance pre-publishing.

![273900623-8f1dbadf-5445-430d-b560-49780206a285](https://github.com/wagtail/wagtail/assets/4907819/20e4e084-1500-46ae-a681-a1e42f5b9694)

![Screenshot 2023-10-10 at 14 00 42](https://github.com/wagtail/wagtail/assets/4907819/56004aae-687c-42c2-b80d-c8e3ea26626c)

<img width="1512" alt="Screenshot 2023-10-10 at 13 40 39" src="https://github.com/wagtail/wagtail/assets/4907819/b4dd4f46-3707-4a2a-856f-cdfcf6baaf6f">

[Figma prototype](https://www.figma.com/proto/h67EsVXdWsfu38WGGxWfpi/Wagtail-CMS?type=design&node-id=13864-11373&viewport=-4573%2C-9678%2C0.5&t=ARpB8DTslaxUJLwN-0&scaling=scale-down-width&starting-point-node-id=13864%3A11373&show-proto-sidebar=1&hide-ui=1)

**Comment by vossisboss** [comment]
I honestly don't find a lot of these tools useful as a content creator. Too often they get in my way and lead to more problems than solutions. For example, MailChimp rolled out a feature this year in which they had a bunch of red exclamation marks that would pop up in my copy every time the AI used in the feature thought my sentences were too long and the only way I could dismiss them was to X out of every single alert. That feature made it take twice as long for me to get a newsletter out the door because the alerts kept blocking my text and annoying me while I was writing. The worst part: It could NOT be switched off.

MailChimp rolled that feature back pretty quickly because they didn't understand that writers write their rough drafts in their email editor. Content creators in Wagtail no doubt do the same thing.So please please please, if we attempt this: Make these features optional and make them something that the editors can control and not need admin permissions to switch on or off.

Also, how are we planning to measure "content quality"?  Content metrics are highly subjective and really cannot be determined universally. Every single audience has different needs and everyone uses different metrics to hit their goals, so I think we should focus more of our energy on giving content people the ability to pull in the metrics that are useful to them like Google Analytics or CTA clicks or page visits. An arbitrary "content quality" thing is really not useful at all because the definition of quality is different for every organization. For example, SEO and accessibility reviews are much more useful for private businesses. Readability matters more to people working in education or government. I think we need to provide people with options for the metrics and measures they want to track.

I don't think this feature would be ideal for Wagtail core unless people can switch the pieces they want on and off or choose what contributes to their content quality score. Most writers are perfectionists with fragile egos. We do not like having low scores. But often scores on "quality" do not correlate with the performance or usefulness of a piece of content. I'd say 99% of writers would find performance metrics more useful for their day to day work than quality. Because if a piece of content has a 100% content score and no one is reading it, then there's still a problem that needs to be solved.

I feel these checkers would be a much better fit for a separate package or a plugin that people can add if they want them or need them. Everyone grades things differently. I think we can get on our soapbox and set the bar higher when it comes to accessibility, but I don't think we should be doing that for "quality" because that is too nebulous to define.

**Reply by thibaudcolas** [reply]
@vossisboss thanks for the feedback! Keep in mind what Ben put together so far is focused on the UI / editor experience. The checks themselves will definitely be configurable and extensible both per project / via plugins, and most likely in the interface for CMS users. There’ll be plenty of opportunity for us to discuss which checks are universally applicable and which aren’t. We do want to have this capability in core, because:

- We don’t want individual projects or plugins to have to reinvent the editor experience to introduce a new check.
- Accessibility checks are mandatory for us to implement as part of meeting the [ATAG 2.0](https://www.w3.org/TR/ATAG20/) standard.

The main driver at this time are those accessibility checks, though we also regularly see projects interested in SEO & sustainability-related checks.

---

> Make these features optional and make them something that the editors can control and not need admin permissions to switch on or off.

I believe we’ve discussed the checks being optional, with accessibility checks on-by-default, and other checks TBC. For the metrics like reading time / words, we’ve not discussed yet.

> Also, how are we planning to measure "content quality"? Content metrics are highly subjective and really cannot be determined universally.

The way we’ve discussed this so far, the one™ "content quality" score would be an aggregate of the scores of specific checks, each with different weights. A perfect "100" score would mean "no issues detected", and from there the score would go down depending on the number of issues. We plan to have universally relevant checks (accessibility, content weight, etc), which would be configurable per project depending on how much checking is appropriate. For example, a site could want a page’s carbon emissions from content to be below a given threshold – with the fundamental checking mechanism built as a package, and the acceptable threshold set either site-wide, or by page type, or by page.

**Reply by vossisboss** [reply]
@thibaudcolas Sounds to me like we're not measuring content quality. Because a piece of writing could pass all those checks and still be terrible and not perform well. For most content creators quality = performance = how many people are actually reading or using this content. I think that by calling this content quality, we're going to introduce confusion and make editors' jobs harder, not easier. I think that if passing checks is what we're aiming for, then we should actually use checkmarks or a pass/fail system similar to when the checks run for a PR submission instead of a score that has no bearing on what content creators consider "quality" to be. I would like to strongly recommend we move away from the scoring idea for now and collect more user research before we proceed with something like that.

**Reply by lb-** [reply]
Maybe we can rename this to content metrics? Content audits, measures, summary. 

Giving this a more generic name may open up for more use cases and a focus on those that are easy to incorporate in core but allows for extensions to build on this. 

Similar to reports, as a mechanism for generic listings of things. 

Metrics, as a mechanism for measuring and adding thresholds for data within your Pages or Snippets.

**Reply by vossisboss** [reply]
Metrics works a little better. It's definitely a more neutral term. It would definitely make it easier for people to extend things. If the intention of the checkers though is to have checks that are used in the process of creating content then I still feel like a checkmark or pass/fail indicator would be more useful for editors.

**Reply by thibaudcolas** [reply]
For what it’s worth, Siteimprove calls this [Content experience](https://www.siteimprove.com/product/content-experience/).

There will definitely be checkmarks and a CI-style pass/fail system for sites that want this kind of protection from quality issues, but scores have the clear advantage that they can summarize quite a complex set of checks. They also offer more options for sites to track possible improvements over time, compared to a binary pass/fail.

> what content creators consider "quality" to be

We want organizations to be able to rely on the CMS so their editors have done a baseline amount of quality checks, regardless of whether individual editors have it on their radar to meet those criteria. It’s certainly not for us to judge whether content performs well, but there are clear, universal quality considerations we can take care of (see what Siteimprove includes as an example).

**Reply by vossisboss** [reply]
@thibaudcolas 

> We want organizations to be able to rely on the CMS so their editors have done a baseline amount of quality checks, regardless of whether individual editors have it on their radar to meet those criteria. I

Organizations care about performance metrics too. And most of them understand that overall content quality incorporates content performance. Calling something a "Quality Score" that is really just a series of QA checks is not accurate and does not reflect the reality of how most content people decide to invest time and resources in improving content. This "Quality Score" would also ultimately not help organizations with limited resources decide what pieces of content to remove, take down, or invest time in refreshing when they've achieved a baseline level of quality, so this score would be of limited use to mature organizations with larger content libraries.

These scores would also have a ceiling and would create more work for content creators because they would have to explain to their stakeholders over and over again why a piece of content with an 100% "Quality Score" still needs improvement and investment. This would absolutely not be helpful to them at all and would make their jobs harder. I'm being pretty adamant about this because I know this would have made my job 1000% harder when I was creating content for large organizations where the attention spans of higher ups are very limited and they often need to be re-educated on what the correct measures for content performance are.

It looks like the Siteimprove example includes performance analytics as a part of their overall score. Ours would not. They have something called "Quality Assurance". That would be a much more appropriate label than "Quality" alone. Most people recognize that QA is connected to checks like spellcheck, etc. "Quality" alone though implies a connection to content performance. If we're going to pursue the scores, then could we at least change the name to "QA Score"? Because if we want content editors to embrace and enjoy Wagtail, we have to show we understand what their needs are and not make their jobs harder.

**Reply by streats** [reply]
Might I suggest "Content integrity" as an alternative to "quality"? If we're mainly focusing on things like accessibility and SEO, we could think of them as the 'structural integrity' of the content and meets certain standards. Though there's a risk "integrity" might be interpreted as challenging the moral/ethical substance of the content itself, I think in context once people see the types of things it's scoring that shouldn't be an issue (and accessibility is sort of a moral/ethical factor anyway).

**Reply by thibaudcolas** [reply]
@streats I think at this stage it’d help to think of examples that are backed by what existing tools do in this space. So we discussed Siteimprove, could also look at e.g. Silktide, Monsido, I’m sure there’s more.

Integrity seems problematic to me for the reason you mentioned. From my experience of it as a concept it also feels like something that’s binary (either something has integrity or it doesn’t), while we’re looking at a UI to suggest improvements that could be very discretionary.

**Comment by allcaps** [comment]
Great! More and more organisations are becoming aware that they should provide accessible content, so the proposed features are very welcome.

I see a "Live analysis" toggle. It appears that the analysis can be switched on and off. 👍

We might consider shipping basic metrics with Wagtail and having a registry in code to add third-party packages. This approach would integrate nicely with Wagtail as a CMS framework. It would also allow analysis packages to mature outside of core. With a registry, integrators could have some control over which panels to display and in what order.

Let's avoid focusing solely on analysis that works exclusively for the English language. For example, German might require 130% more characters for the same text compared to English, thus necessitating different calculation rules.

Have you ever taken a look at https://github.com/Aleksi44/wagtailyoast? It provides the popular Yoast analysis (https://yoast.com/). The strategy of this package involves analyzing the preview. So, the entire process is client-side, and the analysis occurs on the rendered output. I like that strategy.

Keep up the good work! 👏

**Reply by thibaudcolas** [reply]
@allcaps thanks for the feedback Coen!

Yes, the current plan is for most of the analysis to be client-side like `wagtailyoast` does. We see more value in providing results based on full rendered pages than on individual fields. This isn’t set in stone however. 

> Let's avoid focusing solely on analysis that works exclusively for the English language. 

We did discuss this and want to be careful with what we introduce "in core" vs. not in core. It seems like there are a few checks which we could have a universal enough implementation of in core, possibly by allowing projects to define their "target" per site, per page type, per language, or even per page:

- Word count
- Reading time
- Readability checks (TBC)
- Jargon checks (prevents entering words that are banned in the content authors’ style guide)

**Comment by itzomen** [comment]
I love this as this would be extremely helpful in managing and updating content especially those created when Wagtail didn't have the accessibility check ✅.

The following would be a top priority to us:
- As long as, how the content quality check is calculated is clearly stated in the UI, so stakeholders and editors can easily understand it
- An easy way is offered to customize the content quality formula and probably even tweak it directly on the dashboard to assist different set of tasks 
- A way to set the verbosity or quickly select a subset of checks (e.g some people may want it to be more verbose and some may need only a subset of the checks)

Looking forward to this 🎉

**Comment by thibaudcolas** [comment]
For anyone interested in this – I wrote a "stakeholder explainer" blog post, [Looking for sponsorship – Accessibility checks for site administrators](https://wagtail.org/blog/looking-for-sponsorship-accessibility-checks-for-site-administrators/), which ties this a bit more with discussions in #8325 as well. We’re actively looking for sponsorships for this work, particularly on accessibility improvements as there is a clear potential for wide-reaching impact here. Any leads very much appreciated :)

**Reply by cyface** [reply]
I originally talked to folks @ Torchbox about this work a few years ago, and I think the MVP of checks that Thibaud proposes are a good starting point.  And I think the key for me is to provide a framework to allow plug-ins to easily expand this, since extensive expansion of the Wagtail UI in a way that easily survives version upgrades can be tricky.

Specifically, such a framework ideally would:

1) Provide a UI hook to add buttons to initiate content checks.
2) Provide a way to display the results at a page, field, and text-selection level.
3) Allow a "filtering" interface to show/hide the results of checks that collaborators might care about, but you might not; allowing you to focus on the feedback you care about.
4) Allow a way to run these checks _en masse_ and attach the feedback to all the pages; ideally providing a way to send results to collaborators in a CI/CD pipeline sort of way.

I personally was thinking of another sidebar on the right side, along with preview and comments, where checker feedback could be summarized, but also attached _in situ_ to the content it was commenting on.

This is definitely something we'd be excited to spend our unused TMF Wagtail support hours on!

**Reply by thibaudcolas** [reply]
Thank you @cyface! That’s excellent feedback for us.

> I think the key for me is to provide a framework to allow plug-ins to easily expand this

Yes! We started with the UI/UX because it’s the simplest way to demonstrate the concept, I think the next step will be to work a bit on API design / data modelling to make sure we can essentially provide a versatile enough UI that will work well for checks from different tools, of different nature. In a CI/CD pipeline analogy – how do we display "notice" vs. "warning" vs. "error", and what interaction options are there for each (based on what the checker reports, how the site is configured generally, and what user looks at the results)

> I personally was thinking of another sidebar on the right side, along with preview and comments, where checker feedback could be summarized, but also attached in situ to the content it was commenting on.

Could you expand on this? It’s been tricky for us to figure _that_ out because we see people wanting to use the form, and the preview, and seeing the results of comments and checks all at the same time – but there’s only so much space in the UI.

It could be that we bring those automated results within the minimap, so they’re visible at the same time as the preview – or we merge them with comments (with the idea of comments coming in either from automated reviews or humans).

**Reply by cyface** [reply]
Sure!

 So for things like Word Count, I agree that seeing them continually would help; so perhaps my thinking doesn't apply to that.  It might need to get tucked into the header or footer.

 But for contextual suggestions, like "You could link this word to this page", or grammar suggestions, or "this phrase is not allowed by style or legal guidelines", or "it looks like you didn't finish writing this (from AI)", I think a comments-like panel would be ideal so that it could be viewed in context of the suggestion.

 Perhaps you could even just use the comments panel, but allow filtering, "show comments from 1) reviewers, 2) link whisperer tool 3) grammar tool 4) style checker, 5) stock ticker extractor" kind of thing, so people could focus on what they care about from the comments.   One question with this approach would be what to show in the bubble as far as comment count.  

 So for some use cases, I think it would be best to trigger the scan+comment generation explicitly, others should happen on save, and might even prevent the page from saving and / or being published.  For example, not everyone might want grammar suggestions, while style guideline enforcement might be required.
 
 Much like workflow, it would need to be configurable by page type and tree location.
 
 As far as seeing everything at once; many people have fairly wide monitors these days, so I feel like enabling multiple right hand panes might be viable. For example, one pane for preview, and one pane that is task-focused.  Or perhaps make a live-updating preview possible in a separate window, so it could be placed anywhere, like on an iPad to the side of my laptop. :)

**Reply by thibaudcolas** [reply]
Thanks Tim! @benenright what do you think of multiple panels to make the most of screen space on wide screens?

**Comment by lb-** [comment]
## Integration with workflow

I think it would be good to consider how these features would be made available or integrated with the Workflow process.

Viewing stats inside the page editing interface and the dashboard is good but having the information available in context (summarised) for those reviewing content is probably going to be super helpful. 

Workflow rules could even be set up to add/skip steps based on specific metrics thresholds.

**Comment by allcaps** [comment]
I've worked on a couple of websites that have requirements on content block ordering: The intention is that specific page types should contain specific blocks, in a specific order.

Wagtail has ways to validate streamfield blocks. Validation results in errors, preventing the page to be saved. 
The reality is that some content might not be available yet, or there are other good reasons that justify exceptions.

It would be nice to have a _**content quality check**_ to inspect the existence and order of streamfield blocks and _**warn**_ about missing blocks and incorrect order of blocks.

**Comment by lb-** [comment]
## Vocabulary list

I'd like to suggest we pull out a really useful feature from the Value library called 'Vocabulary'. It's a simple list of strings with some metadata for each. 

It's both a way to communicate that some spelling is ok and also enforce specific case-sensitive spelling. This means with one approach you can ensure that things like proper nouns, hyphenated words, brand or organisation specific words are all allowed and enforced. 

I know that we don't want to tackle spelling but we may want to consider the case of validation of proper nouns or hyphenated words. 

https://vale.sh/docs/topics/vocab/

**Comment by thibaudcolas** [comment]
👋 quick progress update, and we could use feedback on what to do next with content checks. In Wagtail 6.2 we implemented [content metrics](https://guide.wagtail.org/en-latest/releases/new-in-wagtail-6-2/#content-metrics) and updates to the Checks panel, thanks to a sponsorship by [The Motley Fool](https://www.fool.co.uk/) (❤️). We also implemented our first custom check, which [flags bad image alt text](https://guide.wagtail.org/en-latest/releases/new-in-wagtail-6-2/#alt-text-accessibility-check).

Those three changes have opened up a lot of possible improvements. We can:

- Add more custom accessibility checks, or customize our existing checks (#12091, #12188)
- Implement other types of checks, such as #12252
- Finally get a better UX to show errors without leaving the page editor, #12187
- Start implementing repair suggestions for common errors, #12260

In addition we also have better support for mapping errors to specific fields coming along (#12045), which means that even if we can’t automate fixes, we could at least point users to the relevant part of the page editor form.

---

So – lots of options. Lots of iterative improvements that will make those checks more capable than ever. I’d be very interested to hear what people in this thread find interesting, or any feedback about anything else you think we should consider.

**Reply by cyface** [reply]
Hi there! As always, extensibility to be able to add our own checks is #1.
The better UX to show issues in the editor rather than the rendered view is also nice.
Cheers!
